hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieWriteHandle.java

```diff
@@ -336,4 +352,51 @@ protected static Option<IndexedRecord> toAvroRecord(HoodieRecord record, Schema
       return Option.empty();
     }
   }
+
+  boolean shouldTrackEventTimeWaterMarkerByPayload(HoodieTableMetaClient metaClient) {
+    // Only for event time ordering mode.
+    if (metaClient.getTableConfig().getRecordMergeMode()
+        != RecordMergeMode.EVENT_TIME_ORDERING) {
+      return false;
+    }
+
+    // Only when `DefaultHoodieRecordPayload` was the payload class.
```

Comment:
for v9 table, we should expect this to be part of write properties. 
only incase of v8 tables, writers will go through payloads and we don't need this fix at all. 
we can simplify this.